### ISSUE TITLE
[Project] Proposal I - SolidJS, Rust / Actix, FaunaDB, Tauri

### ISSUE THREAD

**Issue by wiru** [Tersus-Light/tersus-doc-app#7]
# Intro
I wanted to research as many new and performant frameworks as I could within the time. I think everything bar Rust here is pretty exciting and new. I am not sure whether going for FaunaDB was a huge mistake, but I felt it was perhaps a good way to find the time for Rust to fit into the equation. 

All in all, I believe this proposal will be cost effective, fast and resource efficient, as well as being as secure as possible. Additionally, it takes advantage of some extremely cutting edge technology. 

# Tech Stack
| Technology | Application | Summary |
| ------------- | ------------- | ------------- |
| SolidJS | Frontend Framework | Extremely performant modern framework with true render once components. Scales well. React developers can learn especially quickly. |
| FaunaDB + GraphQL | Resource Server | Several levels of security. Server-less. Globally performant. No upfront cost |
| Rust / Actix-web |  Authentication Server | High security. Steep learning curve. More performant than Golang |
| Tauri | Native Desktop App | Great security. Small bundle sizes. Low memory usage. |

# Architecture
![Architecture](https://user-images.githubusercontent.com/75319621/184340703-e61f57ad-f5e4-4533-99e1-eeea4b1f93a4.jpg)

# Selection Criteria
### Frontend – Solidjs
SolidJS is a relatively new frontend framework which focuses on performance and improving on way reactivity works in other frameworks. It uses the JSX templating syntax which will be familiar to react developers, ensuring a large pool of potential developers who should be able to make the switch to SolidJS easily.

It has a has a [huge approval rating ](https://2021.stateofjs.com/en-US/libraries/front-end-frameworks/)amongst its users as of 2021.  No need to build for optimization due to the compiler (which is larger than Svelte's but also [more sophisticated](https://ryansolid.medium.com/javascript-ui-compilers-comparing-svelte-and-solid-cbcba2120cea))

High performance comes from components truly rendering only once. It is possible to target a specific element in the dom directly and **only** change that element, so no need for diffing. Solid scales very well compared to other compiled frameworks as noted in this [article](https://betterprogramming.pub/the-real-cost-of-ui-components-6d2da4aba205#914b) and specifically [slowdown](https://miro.medium.com/max/1400/1*Z93tkZx3NX_ffzqDPM7Zkw.png) and [memory allocation](https://miro.medium.com/max/700/1*tufgk-XMUFSanlLintKHIA.png) here. 

Solid employs unidirectional data flow, embracing CQRS patterns and immutable interfaces. This helps to improve [performance, scalability and security](https://docs.microsoft.com/en-us/azure/architecture/patterns/cqrs) by **separating read and write** operations to a data store. The compiler will help our project remain less error prone. Solid has been benchmarked with multiple components to see how memory usage is affected at scale, and it appears to perform extremely well. 

The lions share of front end developers are experienced with React, and Solid is based on the same JSX templating style. So it would be relatively easy to find an experienced React developer who could transition to Solid (as more and more are now [doing](https://typeofnan.dev/solid-js-feels-like-what-i-always-wanted-react-to-be/)). 

I had never heard of Solid before looking at @Arcaster42's  proposal. On a personal note, I feel that perhaps I could pick up Svelte more easily than Solid due to my experience with Vue. However, given Reacts stranglehold on the market, I do feel Solid would give me a great chance to become more familiar with Reacts syntax. After having watched a great video by [Rich Harris](https://www.youtube.com/watch?v=AdNJ3fydeao&ab_channel=YouGottaLoveFrontend) (Svelte) about Reactivity and the future of frontend, and contrasting that with a rebuttal to some of his points by [Ryan Carniato](https://ryansolid.medium.com/?source=post_page-----cbcba2120cea--------------------------------) (Solidjs). I personally found myself more often agreeing with Solidjs creator about issues such as slightly larger compile size for increased [performance](https://miro.medium.com/max/1400/1*DGGy6JXWpsl-EYBzvbxZHA.png). I was also very swayed by the arguments surrounding immutable interfaces/unidirectional data flow in terms of creating secure, less error prone, and scalable apps.

### Resource Server – FaunaDB + GraphQL
[FaunaDB](https://www.youtube.com/watch?v=2CipVwISumA&ab_channel=Fireship) is the first server-less database inspired by the [Calvin Paper](http://cs.yale.edu/homes/thomson/publications/calvin-sigmod12.pdf) and ACID compliant.
It supports multi-tenancy for different clusters or user groups. Keeps a history of changes to documents which can be controlled through set time to live (TTL) so as not use too much space. 

Similar scores to Postgres in a Jepsen test, however [Fauna wins overall](https://fauna.com/blog/compare-fauna-vs-postgres) due to extra features which cannot be tested with Postgres. Fauna can deal with relational, graph, temporal and document use cases. 

May not be the cheapest option, but is cost efficient in the sense that it is [free](https://fauna.com/pricing), until we reach a critical mass of users. This could be the safer path with less initial cost outlay before achieving success, especially with no server to maintain. 

Coming back to the Calvin Paper, this cloud based system performs extremely well across distant geographical areas. With remote work across countries on the [increase](https://asia.nikkei.com/Spotlight/Datawatch/Cross-border-online-work-increases-30-amid-COVID-19-pandemic), this could be a big plus for Fauna. Here is the conclusion of the above linked paper provided for convenience.

![Conclusion](https://user-images.githubusercontent.com/75319621/184315821-dd9254f9-5c69-4c06-8e53-1da0ad496949.png)

### Authentication Server – Rust
Rust will take some times for new developers to learn. Certainly more than Golang. However, Golang cannot be as optimized and performant as Rust. This performance though, comes at the cost of increased [complexity and learning time](https://thenewstack.io/rust-vs-go-why-theyre-better-together/).

However, once mastered, Rust is capable of optimizing memory usage sometimes even [2x as well ](https://benchmarksgame-team.pages.debian.net/benchmarksgame/fastest/go-rust.html)as Golang. For just an authentication server though, with a **potential** time saving using a server-less database, the time investment should be more manageable, and allow us to take the harder road in this case. 

Actix-web is a widely popular web framework which has good documentation, runs on stable rust, and is frequently updated. 

On a personal note, I feel that learning about garbage collection will be important for new developers, which Golang would not allow us to learn. 

### Native Desktop Application – Tauri
Tauri – Native Desktop Applicaiton
Tauri is a very secure framework, utilizing [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP), employing [Isolation Patter](https://tauri.app/v1/guides/architecture/inter-process-communication/isolation/)[ns](https://tauri.app/v1/guides/architecture/inter-process-communication/isolation/) as an airlock between back and front end communication. It also only creates binary files as opposed to Electrons ASAR files, which are much [harder to reverse engineer.](https://kylo.tv/how-to-reverse-engineer-electron-ap/) The binary files are configured to only allow specific API functions to be shipped with the app, reducing the possible attack vectors to the backend. 

Memory usage and binary size are vastly improved when compared to electron. This can lead to size reductions of 40x-80x the binary size in the case of a hello world app. 

![Binary Size](https://user-images.githubusercontent.com/75319621/184315239-03d89186-6802-4343-a621-b5fdb6e9ad36.png)

Memory usage is most often between 30%-50% less than that of Electron. 

![Mem Usage](https://user-images.githubusercontent.com/75319621/184315295-8277e743-d301-42be-9745-9ea20fddace1.png)

As Tauri is new, there is a smaller community to learn from, which could be a small concern when looking for support. But whilst, it uses Rust in the backend, minimal experience of Rust is actually required to use Tauri.

# Defense of architecture
I am proposing one server dedicated to authentication. 
The database will be sever-less and use a new cutting edge database tech. Having a cloud based service dedicated to the security of our database provides us with huge geographical coverage that would otherwise cost more for a small project. 
This also frees up some time to allow team members to take on the more daunting task of learning Rust for the authentication server. Which should increase development speed in comparison to taking on multiple new back-end languages. 
Additionally, the flexibility of FaunaDB in a cloud based system essentially gives us the ability to simultaneously use multiple database paradigms all within one (initially) low cost / free service.

**Comment by Arcaster42** [comment]
Hey Will, just wanted to provide some feedback.

Instead of writing out bullet points along with the guidelines, turn it into a more natural conversation. Avoid writing in a presentation-like sense. Keep in mind, this is to learn how to write proposals. In a business setting, there won't be a syllabus, and you'll have to write everything out on your own naturally.

Comments like "fine-grain reactivity" are a little vague. Omitting that and leaving the explanation is good.

Explain as though your target knows the general concepts of what you're discussing. No need to explain what the virtual DOM is or that a compiled language is performant over an interpreted language. This point will be hard as you are coming across some concepts for the first time, so continue to write it out when in doubt, and I'll pass along what a senior engineer should know.

Linking sources is great. When you link images such as your "result1" and "result2", actually write the percentages and link to the source, or post the actual images in the writing. Don't force your reader to click just to figure out what information is behind it.

Very good start and a lot of solid information. Focus on adjusting the presentation of information and cleaning up the explanations.

**Comment by wiru** [comment]
Thank you so much for this feedback. I was a bit stuck on how to present the things I had learnt during the investigation and research phase. 

Your comments are duly noted and I shall endeavor to provide a more polished final product. 

Thanks for the the super feedback. It really means a lot. 

Will

**Comment by Arcaster42** [comment]
I'll add some notes here focusing on the writing.

- In your architecture graph, the client submits a password and receives a token from the auth server, but then passes a token and receives a password from the database? I'm assuming you meant resource instead of password.
- Excellent use of sources, but try to avoid Medium as a source where possible since it's a very casual blog with a lot of misinformation.
- Excellent point on React developers being able to move to Solid.
- Speaking on personal opinion is fine, but speaking on personal scope is a little too informal. "I feel Svelte has a better learning curve" is better than "I feel I could personally learn Svelte faster".
- No need to reference another proposal. Pretend other proposals don't exist.
- Mentioning whitepapers or research papers is typically out of the scope of a proposal. Linking to them is fine. But longer excerpts and details should be left in the source. A source to the ACID concept would be great.
- Awesome, awesome work mentioning Fauna as a serverless DB. That's an innovative pitch.
- Too much information about how Fauna is great. It should be consolidated into one paragraph with appropriate sources and not compared across multiple sections.
- You claim Rust will take longer to learn. Why?
- Great point about trade-offs saving time with the DB to devote more time to the Rust server.
- You talked about learning garbage collection with Rust but not with Go. You overstepped a bit into uncharted territories and accidentally revealed some missing pieces. Rust does not have garbage collection and instead has a concept of ownership and borrowing. Go is a garbage-collected language. Using one or the other doesn't necessarily mean you'll learn about garbage collection.
- Great job mentioning the isolation patterns in Tauri as security is a huge focus for it.
- Great points on performance.

Overall, this is a great first tech proposal and overall a solid proposal. Great balance on simplicity in front-end and innovation on the backend.

**Comment by dawndarkness** [comment]
An enjoyable read for me. Some points I'd like to mention:

- Well researched
- Personally I feel that a few less links would have been a little less distracting
- I think the writing style could be sharpened up a bit, there were some typos and it read a little like a school assignment
- Love the summary at the beginning it was very clear and concise
- You seem to have a good grasp about what is needed for the project

Cheers, bud

**Comment by wiru** [comment]
Hi @Arcaster42 
- Yes, the word password is a typo. It should says resources. 
- Regarding my comment on Garbage collection, my point was simply that we would need to learn more about Garbage Collection using Rust than using Go, as Go does it automatically right? Upon further reading, it appears Rust also seems to GC, but when things go out of scope. So its like a Regional Garbage Collector. I assume this is the concept of ownership that you mentioned before?
- Regarding why is Riust harder. It appears you need a deeper understanding of references and lifetimes, and you must interact more directly with the memory. But I honestly, at this stage I dont know enough to give a solid answer to this past "everyone on the internet says so"
- Thanks for all the other feedback! Duly noted!

**Comment by wiru** [comment]
Hi @dawndarkness 

- Duly noted on the links 
- Yes, I should for sure have more carefully checked for typos, that's a good point.
- Do you have any advice on how to make it sound less like a school assignment? 
- Thank you for the other feedback too!

**Comment by petegriffith** [comment]
Hey hey Will, nice one! This is a very extensive writeup, and I learned a whole lot reading through it last night and again today. 

I noticed a lot of what TJ and Julie mentioned and I don't want to repeat them, so I will focus on one slightly more high-level point. If the goal of the writeup was merely to convince readers that your tech choices are good choices, this checks all the boxes: lots of data to back up your claims, plenty of information on each individual technology. I wonder if it wouldn't be more effective though to focus a bit more on what makes these technologies better/more interesting( /whatever other metric you want to use) _than the other options available._ You briefly touched on Solid vs Svelte, and you definitely mentioned that Solid might be good for React devs. But why didn't you choose React? Or Vue? In other words, you've made a solid argument ( :wink: ) _for_ the technoligies, but perhaps not _as opposed to_ other technologies.

I'm being a bit nitpicky to be honest because Julie and TJ mentioned a lot of things before me, both good and bad, so take it with a grain of salt.

Cheers!
